lightning/src/chain/mod.rs

```diff
@@ -9,6 +9,101 @@
+}
+
+/// An error when accessing the chain via [`Access`].
+///
+/// [`Access`]: trait.Access.html
+#[derive(Clone)]
+pub enum AccessError {
+	/// The requested chain is unknown.
+	UnknownChain,
+
+	/// The requested transaction doesn't exist or hasn't confirmed.
+	UnknownTx,
+}
+
+/// The `Watch` trait defines behavior for watching on-chain activity pertaining to channels as
+/// blocks are connected and disconnected.
+///
+/// Each channel is associated with a [`ChannelMonitor`]. Implementations are responsible for
+/// maintaining a set of monitors such that they can be updated accordingly as channel state changes
+/// and HTLCs are resolved on chain. See method documentation for specific requirements.
+///
+/// TODO: Add documentation about persisting monitors.
+///
+/// If an implementation maintains multiple instances of a channel's monitor (e.g., by using a
+/// watchtower), then it must ensure that updates are applied across all instances. Otherwise, it
+/// could result in a revoked transaction being broadcast, allowing the counterparty to claim all
+/// funds in the channel.
+///
+/// [`ChannelMonitor`]: ../ln/channelmonitor/struct.ChannelMonitor.html
+pub trait Watch: Send + Sync {
+	/// Keys needed by monitors for creating and signing transactions.
+	type Keys: ChannelKeys;
+
+	/// Watches a channel identified by `funding_txo` using `monitor`.
+	///
+	/// Implementations are responsible for watching the chain for the funding transaction along
+	/// with any spends of outputs returned by [`get_outputs_to_watch`]. In practice, this means
+	/// calling [`block_connected`] and [`block_disconnected`] on the monitor.
+	///
+	/// [`get_outputs_to_watch`]: ../ln/channelmonitor/struct.ChannelMonitor.html#method.get_outputs_to_watch
+	/// [`block_connected`]: ../ln/channelmonitor/struct.ChannelMonitor.html#method.block_connected
+	/// [`block_disconnected`]: ../ln/channelmonitor/struct.ChannelMonitor.html#method.block_disconnected
+	fn watch_channel(&self, funding_txo: OutPoint, monitor: ChannelMonitor<Self::Keys>) -> Result<(), ChannelMonitorUpdateErr>;
+
+	/// Updates a channel identified by `funding_txo` by applying `update` to its monitor.
+	///
+	/// Implementations must call [`update_monitor`] with the given update. See
+	/// [`ChannelMonitorUpdateErr`] for invariants around returning an error.
+	///
+	/// [`update_monitor`]: ../ln/channelmonitor/struct.ChannelMonitor.html#method.update_monitor
+	/// [`ChannelMonitorUpdateErr`]: ../ln/channelmonitor/enum.ChannelMonitorUpdateErr.html
+	fn update_channel(&self, funding_txo: OutPoint, update: ChannelMonitorUpdate) -> Result<(), ChannelMonitorUpdateErr>;
+
+	/// Returns any monitor events since the last call. Subsequent calls must only return new
+	/// events.
+	fn release_pending_monitor_events(&self) -> Vec<MonitorEvent>;
+}
+
+/// The `Notify` trait defines behavior for indicating chain activity of interest pertaining to
```

Comment:
I'm not sure about the "Notify" name, but, more broadly, should we endeavor to have ChainMonitor::block_connected just return the to-be-registered bits instead? I'm still not convinced its not a better design by avoiding reentrancy, but maybe I forgot some part of the conversation?